### PR TITLE
feat: add balance loading backend infrastructure for automatic refund funding

### PR DESCRIPTION
**Pull request by yashranaway**
# PR 1: Balance Loading Backend Foundation #1594

I have split the work into three PRs to keep things clean and manageable  one for the backend setup, one for the APIs, and one for the frontend so each part can be reviewed and tested separately. 

## What

Implemented backend infrastructure for balance loading feature that allows sellers to connect credit cards and automatically load funds to their Gumroad balance when needed for refunds. Created database tables, models, services, background jobs, and comprehensive test suite following Gumroad's architectural patterns.

## Why

Sellers frequently need to issue refunds but lack sufficient balance after payouts have been processed. Currently, they must wait for new sales or manually transfer funds, delaying customer refunds and creating support friction. This feature enables sellers to attach a credit card that automatically charges when their balance is insufficient for a refund, ensuring immediate refund processing and better customer experience.

Implements feature request for instant refund capability with automatic balance loading.

## Changes

**Database Schema** (2 migrations)
- Created `balance_load_credit_cards` table to store Stripe payment methods with soft delete support
- Created `balance_loads` table to track charge attempts and link to refunds
- Added proper indexing for performance (external_id, stripe IDs, user lookups)
- Added foreign key constraints and unique constraints for data integrity

**Models** 
- [BalanceLoadCreditCard](cci:2://file:///Users/adityagarud/gumroad/app/models/balance_load_credit_card.rb:2:0-44:3) - Manages credit cards with ExternalId, Deletable concerns
  - Validates card details (last4, brand, expiration)
  - Enforces one default card per user
  - Auto-unsets other defaults when setting new default
  - Provides expiration checking and display formatting
- [BalanceLoad](cci:2://file:///Users/adityagarud/gumroad/app/models/balance_load.rb:2:0-52:3) - Tracks charge lifecycle with PaperTrail audit trail
  - State machine: `pending` → `successful` / `failed`
  - Links to refund for audit trail
  - Converts cents to dollars for display
- `Refund::BalanceLoadIntegration` concern - Integrates with refund flow
  - Checks seller balance before refund
  - Auto-loads insufficient amount
  - Respects `balance_load_for_refunds` feature flag

**Services** 
- [BalanceLoading::PaymentMethodService](cci:2://file:///Users/adityagarud/gumroad/app/services/balance_load/payment_method_service.rb:3:2-54:5) - Stripe payment method management
  - Attaches payment methods to Stripe customers
  - Handles soft delete vs hard delete based on usage
  - Sets default payment method
  - Comprehensive error handling for Stripe failures
- [BalanceLoading::ChargeService](cci:2://file:///Users/adityagarud/gumroad/app/services/balance_load/charge_service.rb:3:2-84:5) - Card charging logic
  - Validates minimum charge amount ($1.00)
  - Creates pending BalanceLoad record
  - Enqueues async processing job
  - Creates Balance record on success
  - Handles card errors and Stripe failures

**Background Jobs** 
- [BalanceLoading::ProcessChargeJob](cci:2://file:///Users/adityagarud/gumroad/app/sidekiq/balance_load/process_charge_job.rb:3:2-24:5) - Async charge processing
  - Queue: `default` (time-sensitive for refunds)
  - Retry: 3 attempts with `lock: :until_executed`
  - Notifies Bugsnag on failures
  - Updates BalanceLoad status

**Tests**
- Model specs: Validations, associations, scopes, state transitions
- Service specs: Stripe integration (mocked), error handling, edge cases
- Job specs: Async processing, retry logic, error notification
- Integration specs: Refund flow with feature flag, balance checking
- Factory specs: Traits for different card states (default, expired, deleted)

**User Model Integration**
- Added associations: `balance_load_credit_cards`, `balance_loads`
- Added scoped associations: `alive_balance_load_credit_cards`, `default_balance_load_credit_card`

## Technical Decisions

**Why `BalanceLoading::` namespace?** 
Avoided namespace conflict with [BalanceLoad](cci:2://file:///Users/adityagarud/gumroad/app/models/balance_load.rb:2:0-52:3) model. Gumroad pattern uses `Purchase::` for purchase-related services, so `BalanceLoading::` follows this convention.

**Why separate `balance_load_credit_cards` table?**
Separates payment method management from charge history. Allows soft delete of cards while preserving charge records. Follows Stripe's separation of PaymentMethod and Charge resources.

**Why async job processing?**
Stripe API calls can take 2-5 seconds. Async processing prevents blocking refund UI and allows retry logic for transient failures.

**Why `default` queue instead of `low`?**
Balance loading is time-sensitive when triggered by refunds. Customers expect immediate refund processing, so `default` queue ensures timely execution.

**Why feature flag?**
Allows gradual rollout, A/B testing, and quick disable if issues arise. Can enable for specific high-volume sellers first.

## Testing

All related model, service, and integration tests have been added and successfully pass. These cover validation rules, Stripe API interactions, refund flows, feature flag behavior, and key edge cases such as expired cards, insufficient balance, and idempotent charge handling.

## AI Disclosure

- **Model**: Claude 3.5 Sonnet via Cursor
- **Used for**: Code generation, test writing, following Gumroad architectural patterns (ExternalId, Deletable, PaperTrail, service objects, Sidekiq jobs)
- **All code manually reviewed** 

## Live Stream Disclosures

I have watched both of the live streams

### CODE REVIEW

File: app/models/concerns/refund/balance_load_integration.rb
Comment (yashranaway) [inline app/models/concerns/refund/balance_load_integration.rb:1]: Stores the seller’s credit cards for automatic balance loading, supports multiple cards per user with a single default, and soft deletes cards once they’ve been used for charges.

File: app/models/balance_load.rb
Comment (yashranaway) [inline app/models/balance_load.rb:1]: Tracks credit card charges used to load the seller’s balance, transitions through states from pending to successful or failed, and links to a refund record when triggered by insufficient balance.

File: app/models/balance_load_credit_card.rb
Comment (yashranaway) [inline app/models/balance_load_credit_card.rb:1]: Checks the seller’s balance before processing a refund and automatically loads funds if required, running only when the `balance_load_for_refunds` feature flag is enabled.

File: app/services/balance_loading/payment_method_service.rb
Comment (yashranaway) [inline app/services/balance_loading/payment_method_service.rb:1]: Manages Stripe payment methods used for balance loading, handling card attachment and detachment as well as default card selection.

File: app/services/balance_loading/charge_service.rb
Comment (yashranaway) [inline app/services/balance_loading/charge_service.rb:1]: Charges the seller’s credit card to load their balance, creates a pending record, and enqueues an asynchronous job, with a minimum charge amount of $1.00.

File: app/sidekiq/balance_loading/process_charge_job.rb
Comment (yashranaway) [inline app/sidekiq/balance_loading/process_charge_job.rb:1]: Processes balance load charges asynchronously, retries up to three times on failure, and reports any errors to Bugsnag.

File: db/migrate/20251022131704_create_balance_load_credit_cards.rb
Comment (yashranaway) [inline db/migrate/20251022131704_create_balance_load_credit_cards.rb:1]: Stores Stripe payment methods used for balance loading, tracking card details, expiration, and default status.

File: db/migrate/20251022131710_create_balance_loads.rb
Comment (yashranaway) [inline db/migrate/20251022131710_create_balance_loads.rb:1]: Tracks charge attempts and their status, linking each charge to a credit card and optionally to a refund.

File: app/models/user.rb [hunk 1]
```diff
@@ -122,6 +122,10 @@ class User < ApplicationRecord
   has_many :community_chat_messages, dependent: :destroy
   has_many :last_read_community_chat_messages, dependent: :destroy
   has_many :community_notification_settings, dependent: :destroy
+  has_many :balance_load_credit_cards, dependent: :destroy
+  has_many :alive_balance_load_credit_cards, -> { alive }, class_name: "BalanceLoadCreditCard"
+  has_one :default_balance_load_credit_card, -> { alive.where(is_default: true) }, class_name: "BalanceLoadCreditCard"
+  has_many :balance_loads, dependent: :destroy
```
Comment (yashranaway) [inline app/models/user.rb:128]: Defines balance loading associations used for automatic refund funding.

File: app/models/refund.rb [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
-  include JsonData, FlagShihTzu
+  include JsonData, FlagShihTzu, Refund::BalanceLoadIntegration
 
   belongs_to :user, foreign_key: :refunding_user_id, optional: true
   belongs_to :purchase
   belongs_to :product, class_name: "Link", foreign_key: :link_id
   belongs_to :seller, class_name: "User"
   has_many :balance_transactions
   has_one :credit
+  has_one :balance_load
```
Comment (yashranaway) [inline app/models/refund.rb:14]: Links to a balance load record when a refund triggers automatic funding.

**Comment by yashranaway** [comment]
@michellelarney Could you please review my approach and let me know if any adjustments or improvements are needed before moving on to the next PRs.

**Comment by michellelarney** [comment]
@slavingia Can review.

**Comment by slavingia** [comment]
Needs to be in one PR with end to end tests and a comprehensive video.
